File: #0042.trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {

        int size = height.size();
        if (size == 0) {
            return 0;
        }

        int i;
        int left[size];
        int right[size];
        left[0] = height[0];
        for(i = 1; i < size; i++) {
            left[i] = std::max(left[i - 1], height[i]);
        }
        
        for(i = size - 2; i >= 0; i--) {
            right[i] = std::max(right[i + 1], height[i]);
        }

        int result = 0;
        for (i = 1; i <= size-2; i++) {
            result += (std::min(left[i], right[i]) - height[i]);
        }
        return result;
    }
};

File: #0042.trapping-rain-water.js
/**
 * @param {number[]} height
 * @return {number}
 */
var trap = function(height) {
    let i = 0;
    let j = height.length - 1;
    let left = 0;
    let right = 0;
    let result = 0;
    while (i < j) {
        if (height[i] < height[j]) {
            left = Math.max(left, height[i]);
            result += left - height[i++];
        } else {
            right = Math.max(right, height[j]);
            result += right - height[j--];
        }
    }
    return result;
};
